pkg/server/singleprocess/service_ui_pipeline.go

```diff
@@ -0,0 +1,81 @@
+package singleprocess
+
+import (
+	"context"
+
+	"github.com/hashicorp/go-hclog"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	pb "github.com/hashicorp/waypoint/pkg/server/gen"
+	"github.com/hashicorp/waypoint/pkg/server/hcerr"
+	serverptypes "github.com/hashicorp/waypoint/pkg/server/ptypes"
+)
+
+func (s *Service) UI_ListPipelines(
+	ctx context.Context,
+	req *pb.UI_ListPipelinesRequest,
+) (*pb.UI_ListPipelinesResponse, error) {
+	if err := serverptypes.ValidateUIListPipelinesRequest(req); err != nil {
+		return nil, err
+	}
+
+	// Create uninitialized array of pipeline bundles
+	var allPipelines []*pb.UI_PipelineBundle
+
+	// Get list of all pipelines
+	pipelineListResponse, err := s.state(ctx).PipelineList(ctx, req.Project)
+	if err != nil {
+		return nil, hcerr.Externalize(
+			hclog.FromContext(ctx),
+			err,
+			"error listing piplines",
+		)
+	}
+
+	// Create bundles
+	for _, pipeline := range pipelineListResponse {
+		ref := &pb.Ref_Pipeline{
+			Ref: &pb.Ref_Pipeline_Id{
+				Id: pipeline.Id,
+			},
+		}
+		// Get total runs for the pipeline
+		pipelineRunListResponse, err := s.state(ctx).PipelineRunList(ctx, ref)
```

Comment:
I can imagine that this list will be very long for some pipelines, and will eventually make this response too large, even if we respect the request pagination!

Might be nice to limit this to something like the 10 most recent runs (or whatever the UI needs), and document that somewhere. Not sure if that state function supports a limit - if it does, use that - but otherwise you can throw out everything but the 10 most recent runs from the list.

Reply:
We are only pulling all of these to get the total count, and not actually returning all the runs themselves. This is an interim solution. We will eventually have the count as part of the state function.

Reply:
One solution for getting total runs could be instead grabbing the sequence number from `PipelineRunGetLatest` instead of counting the total list. The sequence number on a run should match total runs, so the latest sequence should be the same as total runs?

Reply:
@briancain oooh, that’s a very neat idea!